### PR TITLE
Added Pytorch model: Inception V3, MobileNet V2, Resnet-50; and PyTorch-to-ONNX conversion tool

### PR DESCRIPTION
**Pull request by eizamaliev**
Added models, including MO and Pytorch-to-ONNX conversion parameters.

### CODE REVIEW

**Comment by eizamaliev** [comment]
@IRDonch  please review

**Comment by eizamaliev** [comment]
cc @kchechil

File: tools/downloader/list_topologies.yml [hunk 1]
```diff
@@ -2443,3 +2443,122 @@ topologies:
...
+    framework: pytorch
+    pytorch_to_onnx:
+      - --model-name=mobilenet-v2
+      - --model-path=classification/mobilenet/v2/pytorch
+      - --weights=classification/mobilenet/v2/pytorch/mobilenet-v2.pth.tar
+      - --input-shape=1,3,224,224
+      - --output-file=classification/mobilenet/v2/pytorch/mobilenet-v2.onnx
+      - --input-names=data
+      - --output-names=prob
+    model_optimizer_args:
```
[the remark this answers is not in the document]
Comment (eizamaliev) [reply tools/downloader/list_topologies.yml]: see #154

**Comment by eizamaliev** [comment]
@IRDonch please review

**Comment by IRDonch** [comment]
There's the same sort of weirdness going on here as in #154. Please rebase.

**Comment by eizamaliev** [comment]
@IRDonch please review

**Comment by eizamaliev** [comment]
@IRDonch please review and merge

**Comment by IRDonch** [comment]
OK, so normally I'm all about splitting changes, but in this case I think you should combine this PR with !154, because then we can use CI to test that the conversion actually works (testing the two PRs separately won't produce any interesting results).

File: tools/downloader/common.py [hunk 2]
```diff
@@ -26,7 +26,7 @@
 DOWNLOAD_TIMEOUT = 5 * 60
 
 # make sure to update the documentation if you modify these
-KNOWN_FRAMEWORKS = {'caffe', 'dldt', 'mxnet', 'tf'}
+KNOWN_FRAMEWORKS = {'caffe', 'dldt', 'mxnet', 'tf', 'pytorch'}
```
Comment (IRDonch) [inline tools/downloader/common.py]: Keep them in alphabetical order, please. You also need to update the README correspondingly.

File: tools/downloader/list_topologies.yml [hunk 3]
```diff
@@ -3233,3 +3233,125 @@ topologies:
...
+          $type: google_drive
+          id: 1jlto6HRVD3ipNkAl1lNhDbkBp7HylaqR
+    output: "classification/mobilenet/v2/pytorch/"
+    framework: pytorch
+    pytorch_to_onnx:
+      - --model-name=mobilenet-v2
+      - --model-path=classification/mobilenet/v2/pytorch
+      - --weights=classification/mobilenet/v2/pytorch/mobilenet-v2.pth.tar
+      - --input-shape=1,3,224,224
+      - --output-file=classification/mobilenet/v2/pytorch/mobilenet-v2.onnx
```
Comment (IRDonch) [inline tools/downloader/list_topologies.yml]: I don't think the output file name should be specified here, because the converter should be able to derive it automatically based on the model name and subdirectory - much like it does for Model Optimizer.

File: tools/downloader/list_topologies.yml [hunk 4]
```diff
@@ -3233,3 +3233,125 @@ topologies:
...
+      in RGB order. 
+
+      The model output is typical object classifier for the 1000 different classifications
+      matching with those in the ImageNet database.
+    task_type: "classification"
+    files:
+      - name: MobileNetV2.py
+        sha256: 7071e01e07e5f1b6127a2628d1be9c65cc20edfa745f4f4eb6354e7b27f5b444
+        source: https://raw.githubusercontent.com/tonylins/pytorch-mobilenet-v2/fa86a85e4026caa217fe8d62f370848195cedf4b/MobileNetV2.py
+      - name: mobilenet-v2.pth.tar
```
Comment (IRDonch) [inline tools/downloader/list_topologies.yml]: This doesn't seem to actually be a tar file, so to avoid confusion, I think you should remove the `.tar` extension.

File: tools/downloader/list_topologies.yml [hunk 5]
```diff
@@ -3233,3 +3233,125 @@ topologies:
...
+      - --output-names=prob
+    model_optimizer_args:
+      - --model_name=mobilenet-v2
+      - --framework=onnx
+      - --data_type=FP32
+      - --input=data
+      - --mean_values=data[123.675,116.28,103.53]
+      - --scale_values=data[58.624,57.12,57.375]
+      - --output=prob
+      - --input_model=$dl_dir/mobilenet-v2.onnx
```
Comment (IRDonch) [inline tools/downloader/list_topologies.yml]: It is now possible to specify a separate directory for converter output, so you can't use `dl_dir` here. You'll need to add another variable, e.g. `conv_dir`.

Comment (eizamaliev) [reply tools/downloader/list_topologies.yml]: All models (in pth and onnx formats) will be in same, `dl_dir`, directory

File: tools/downloader/common.py (hunk 2 above)
Comment (IRDonch) [reply tools/downloader/common.py]: By "correspondingly" I meant that there's a list of frameworks in the part that talks about Info Dumper's output, which needs to be updated in the same way.

File: tools/downloader/list_topologies.yml (hunk 5 above)
Comment (IRDonch) [reply tools/downloader/list_topologies.yml]: Well, you shouldn't put the ONNX model in `dl_dir` - it's not something that we download.

**Comment by IRDonch** [comment]
Jenkins: test this

**Comment by IRDonch** [comment]
Jenkins: test this please

File: tools/downloader/list_topologies.yml [hunk 6]
```diff
@@ -3286,3 +3286,123 @@ topologies:
...
+      - --output=prob
+      - --input_model=$conv_dir/googlenet-v3.onnx
+    license: https://raw.githubusercontent.com/pytorch/vision/master/LICENSE
+
+  - name: "resnet-50-pytorch"
+    description: >-
+      Resnet 50 is image classification model pretrained on ImageNet dataset. This
+      is PyTorch implementation based on architecture described in paper Deep Residual
+      Learning for Image Recognition" <https://arxiv.org/pdf/1512.03385> in TorchVision
+      package <https://github.com/pytorch/vision>
```
Comment (IRDonch) [inline tools/downloader/list_topologies.yml]: The description doesn't seem to match the original Markdown text - there's a missing quote and no period at the end.

File: tools/downloader/pytorch_to_onnx.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+def parse_args():
+    """Parse input arguments"""
+
+    parser = argparse.ArgumentParser(description='Conversion script for pretrained models from PyTorch to ONNX')
+    parser.add_argument('--model-name', type=str, required=True,
+                        help='Model name to convert.'
+                             ' One of the ' + ', '.join(SUPPORTED_MODELS) + ' is supported')
+    parser.add_argument('--weights', type=str, required=True,
+                        help='Path to the PyTorch pretrained weights')
+    parser.add_argument('--input-shape', metavar='INPUT_DIM', type=positive_int_arg, required=True,
```
[the remark this answers is not in the document]
Comment (eizamaliev) [reply tools/downloader/pytorch_to_onnx.py]: No

File: tools/downloader/pytorch_models_conversion.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
+# Conversion pretrained models from PyTorch to ONNX
```
[the remark this answers is not in the document]
Comment (eizamaliev) [reply tools/downloader/pytorch_models_conversion.md]: Ok, prepare to release this featurelater, but if you insist, I'll do it in this patch.
So, the main issue is that any model can be added to conversion without any changes in script. I tested it on few models and it works.

[the remark this answers is not in the document]
Comment (IRDonch) [reply tools/downloader/pytorch_models_conversion.md]: > Ok, prepare to release this featurelater, but if you insist, I'll do it in this patch.

Wait, what feature? I'm not asking for any new features. I'm actually suggesting the opposite - I don't think the conversion script should be a feature by itself. IMO, it should just be an implementation detail of the model converter.

File: tools/downloader/pytorch_to_onnx.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (IRDonch) [reply tools/downloader/pytorch_to_onnx.py]: Alright, if you say so.

File: tools/downloader/pytorch_models_conversion.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (eizamaliev) [reply tools/downloader/pytorch_models_conversion.md]: Yes, but you pointed on very weak part of converter. So, with your help I fixed it, making it able to solve more common problem.

**Comment by eizamaliev** [comment]
@IRDonch please add `pytorch_requirements.in` to pipeline

File: tools/downloader/pytorch_models_conversion.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (IRDonch) [reply tools/downloader/pytorch_models_conversion.md]: OK, I like what you did wrt arbitrary model support, but I think we should still keep this script internal for now. I don't like the idea of exposing it right after we add it, because as we add more models, we might want to change it in an incompatible way. Let's give it some time to mature.

If you agree, then please remove the separate README and add any relevant parts to the downloader README. If not, then let me know why. :-)

File: tools/downloader/converter.py [hunk 9]
```diff
@@ -38,6 +38,18 @@ def quote_windows(arg):
...
     quote_arg = shlex.quote
 
+
+def convert_to_onnx(topology, output_dir, args):
+    print('========= {}Converting {} to ONNX'.format('(DRY RUN) ' if args.dry_run else '', topology.name))
+    pytorch_to_onnx_args = [
+        string.Template(arg).substitute(conv_dir=output_dir / topology.subdirectory,
+                                        dl_dir=args.download_dir / topology.subdirectory)
+        for arg in topology.pytorch_to_onnx_args]
+    cmd = [str(args.python), 'pytorch_to_onnx.py', *pytorch_to_onnx_args]
```
[the remark this answers is not in the document]
Comment (IRDonch) [reply tools/downloader/converter.py]: No, what you didn't isn't going to help the matters. You need to use `__file__` to determine the location of `converter.py` and find `pytorch_to_onnx.py` relative to that.

File: tools/downloader/pytorch_requirements.in [hunk 10]
```diff
@@ -0,0 +1,4 @@
+numpy
+torch
+torchvision
+onnx
```
Comment (IRDonch) [inline tools/downloader/pytorch_requirements.in]: A few things need to be done here:

* sort the package names;
* rename the file to `requirements-pytorch.in`;
* replace any mentions of specific dependencies in the documentation with a reference to this file;
* use `pip-compile` to build `ci/requirements-conversion.txt` from a combination of this and MO's `requirements.txt`;
* remove `ci/requirements-mo.txt` (I will amend the pipeline to instead use `ci/requirements-conversion.txt` when it's available).

File: tools/downloader/pytorch_requirements.in [hunk 11]
```diff
@@ -0,0 +1,4 @@
+numpy
```
Comment (IRDonch) [inline tools/downloader/pytorch_requirements.in]: I don't think you actually use `numpy` in your script at all.

**Comment by IRDonch** [comment]
Here's a test build with a draft version of the pipeline: http://134.191.240.124/job/omz/job/precommit/job/downloader_ubuntu16/502/console

Note that one conversion failed.

File: tools/downloader/README.md [hunk 12]
```diff
@@ -81,7 +88,8 @@ The basic usage is to run the script like this:
 ```
 
 This will convert all models into the Inference Engine IR format. Models that
-were originally in that format are ignored.
+were originally in that format are ignored. If model in PyTorch's format they will be
```
Comment (IRDonch) [inline tools/downloader/README.md]: You have mismatched tenses here: singular "model" in the if clause and plural "they" in the main clause. Also, there's no verb in the if clause.

File: tools/downloader/requirements-pytorch.in [hunk 13]
```diff
@@ -0,0 +1,4 @@
+onnx
+scipy
```
Comment (IRDonch) [inline tools/downloader/requirements-pytorch.in]: Could you add a comment about why `scipy` is required?

File: models/public/mobilenet-v2-pytorch/mobilenet-v2-pytorch.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+### Converted model
+
+Image, name - `data`,  shape - `1,3,224,224`, format is `B,C,H,W` where:
+
+- `B` - batch size
+- `C` - channel
+- `H` - height
+- `W` - width
+
+Channel order is `RGB`
```
Comment (IRDonch) [inline models/public/mobilenet-v2-pytorch/mobilenet-v2-pytorch.md]: Wouldn't it make sense to reverse the channels during conversion, given that we already do that for most (possibly all) other models?

File: ci/requirements-conversion.txt [hunk 15]
```diff
@@ -17,14 +17,17 @@ mxnet==1.3.1
 networkx==2.3
 numpy==1.14.6
 onnx==1.5.0
+pillow==6.1.0             # via torchvision
 protobuf==3.6.1
 requests==2.22.0          # via mxnet
-six==1.12.0               # via absl-py, grpcio, h5py, keras-preprocessing, mock, onnx, protobuf, tensorboard, tensorflow, tensorflow-estimator, test-generator
```
Comment (IRDonch) [inline ci/requirements-conversion.txt]: Why did you remove `six`?

Comment (eizamaliev) [reply ci/requirements-conversion.txt]: Merge error

File: tools/downloader/requirements-pytorch.in (hunk 13 above)
Comment (eizamaliev) [reply tools/downloader/requirements-pytorch.in]: This is torchvision's requirement, at least one model (inception/googlenet v3) use it

File: models/public/mobilenet-v2-pytorch/mobilenet-v2-pytorch.md (hunk 14 above)
Comment (eizamaliev) [reply models/public/mobilenet-v2-pytorch/mobilenet-v2-pytorch.md]: Because we reverse channels only if they are in `BGR` order. So, all public converted models have `RGB` input. This is a convention for public models, as I know.

File: tools/downloader/requirements-pytorch.in (hunk 13 above)
Comment (IRDonch) [reply tools/downloader/requirements-pytorch.in]: Yeah, _I_ know that. But people who read this later won't. That's why I'm asking you to add a comment.

File: models/public/mobilenet-v2-pytorch/mobilenet-v2-pytorch.md (hunk 14 above)
Comment (IRDonch) [reply models/public/mobilenet-v2-pytorch/mobilenet-v2-pytorch.md]: Are you sure about this? I randomly opened a few models, and all of them have BGR input in the converted form (either because they already had it originally, or because the channels were reversed during conversion).

File: tools/downloader/README.md (hunk 12 above)
Comment (IRDonch) [reply tools/downloader/README.md]: You fixed the first error, but not the second one.

File: tools/downloader/list_topologies.yml (hunk 3 above)
Comment (eizamaliev) [reply tools/downloader/list_topologies.yml]: Could you resolve this issue?

File: tools/downloader/README.md (hunk 12 above)
Comment (eizamaliev) [reply tools/downloader/README.md]: Fixed. Resolve?

File: tools/downloader/list_topologies.yml (hunk 3 above)
Comment (IRDonch) [reply tools/downloader/list_topologies.yml]: No, because you haven't done anything about it.

Comment (eizamaliev) [reply tools/downloader/list_topologies.yml]: Sorry, but I don't see any sense in this change, except automation. Which advantage could we achieve with it?
Could you explain me your point?

Comment (IRDonch) [reply tools/downloader/list_topologies.yml]: We know that for any model we'll need to pass an `--output-file` option to the ONNX converter. We also know that the value for said option will always be `<conv_dir>/<model_name>.onnx`. Specifying this option manually is thus of no benefit and serves only to invite mistakes. It should instead be automatically supplied by the Model Converter.

Comment (eizamaliev) [reply tools/downloader/list_topologies.yml]: It makes sense.
But we must go further in that case and get rid of almost all mo args, like `--framework`, `--input-model` (or any `--input_*`), `--data_type` with same reason. Would you do this?

Comment (IRDonch) [reply tools/downloader/list_topologies.yml]: I was indeed planning on getting rid of `--framework` and `--data_type` (the latter I would have to remove anyway to add support for multiple output precisions). `--input_*` cannot be removed right now, because there's no way to automatically derive the values of these options.

Comment (eizamaliev) [reply tools/downloader/list_topologies.yml]: I cannot agree about `--input_*` with you. I randomly picked up some models and most of them have keys like `--input_*=$dl_dir/<model_name>.<format>`. Only archived models make an exception. So, I supposed that these data are derivable as well. 
In this case, we can indeed lower the tendency for mistakes in configuration. But if changes will be only in part, which you suggest, benefit would turn low. In fact, it would cause misunderstanding in "where is onnx model come from?" for MO argument `--input-model`.

Comment (IRDonch) [reply tools/downloader/list_topologies.yml]: > In fact, it would cause misunderstanding in "where is onnx model come from?" for MO argument --input-model.

OK, that's a fair point. Let keep it as-is for now, then.
